src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  concatToEndOfState: function(propsToConcat, callback) {
+    //throw error if propsToConcat is undefined or not an object
+    invariantArgCheck(propsToConcat);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToConcat, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that currentStateAtKey is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that newProps at this key is an array
+      invariantArrayCheck(newPropsAtKey);
+      //concat newPropsAtKey to the end of currentStateAtKey
+      newState[key] = currentStateAtKey.concat(newPropsAtKey);
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //concatToFrontOfState FUNCTION: concat the arrays of values in the deepest keys on the passed in object to the front of the corresponding arrays in the deepest keys in the current state. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToConcat OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  concatToFrontOfState: function(propsToConcat, callback) {
+    //throw error if propsToConcat is undefined or not an object
+    invariantArgCheck(propsToConcat);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToConcat, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that currentStateAtKey is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that newProps at this key is an array
+      invariantArrayCheck(newPropsAtKey);
+      //concat newPropsAtKey to the front of currentStateAtKey
+      newState[key] = newPropsAtKey.concat(currentStateAtKey);
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //resetState FUNCTION: replace the current state with the resulting value of getInitialState
+  //@param callback FUNCTION: optional callback that will be executed once replaceState is completed and the component is re-rendered
+  resetState : function (callback) {
+    //make a shallow copy of the resulting value of getInitialState
+    // console.log(this.getInitialState);
+    var newState = assign({}, this.getInitialState());
+    //replace current state with the updated values and a callback (if provided), triggering a re-render
+    this.replaceState(newState, callback);
```

Comment:
Don't use `replaceState` here.  Instead just have function do:

this.setState(this.getInitialState());

then add another function called hardResetState which calls this.replaceState(this.getInitialState());